### ISSUE TITLE
Compilation with Clang fails, va_printf undeclared

### ISSUE THREAD

**Issue by markuspetermann** [thrust/thrust#1032]
Trying to compile the following example:

```C++
#include <thrust/device_vector.h>

int main(void)
{
    thrust::device_vector<float> A(1);
    return 0;
}
```

with Clang 8 or Clang 10 yields the following error:

```
/usr/local/cuda-10.0/include/thrust/system/cuda/detail/core/agent_launcher.h:557:11: error: use of undeclared identifier 'va_printf'

/usr/local/cuda-10.0/include/thrust/system/cuda/detail/cub/device/dispatch/../../util_debug.cuh:134:42: note: expanded from macro '_CubLog'
            #define _CubLog(format, ...) va_printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, __VA_ARGS__);

...
```

Compiling with nvcc works.

Checking the code after pre-processing shows that the `va_printf()` function indeed lives under `thrust::cuda_cub::cub` while in `agent_launcher.h` it is called from within `thrust::cuda_cub::core`.

**Comment by j-petit** [comment]
+1, tested with clang 8 under Ubuntu 18.04.

**Comment by j-petit** [comment]
Dirty workaround for me:
```
#define THRUST_CUB_NS_PREFIX namespace thrust::cuda_cub {
#define THRUST_CUB_NS_POSTFIX }

#include <thrust/system/cuda/detail/cub/util_debug.cuh>

using namespace thrust::cuda_cub::cub;
#include <thrust/device_ptr.h>
#include <thrust/reduce.h>
```

**Comment by jngrad** [comment]
The `va_printf` compiler error is reproducible on Clang++ 9.0 with CUDA 10.0 and thrust 1.9.3 when compiling our software ESPResSo. Applying the solution by @j-petit removed the compiler error in ESPResSo, with all GPU tests passing:

```diff
--- thrust/system/cuda/config.h	2020-04-09 17:41:08.164278955 +0200
+++ thrust/system/cuda/config.h	2020-04-09 17:42:30.793338972 +0200
@@ -77,6 +77,8 @@
 #define THRUST_CUB_NS_PREFIX namespace thrust {   namespace cuda_cub {
 #define THRUST_CUB_NS_POSTFIX }  }
 
+#include <thrust/system/cuda/detail/cub/util_debug.cuh>
+using namespace thrust::cuda_cub::cub;
 
 #ifndef END_NS_THRUST
 #define END_NS_THRUST }
```

The underlying issue comes from an old fix in the CUB dependency ([thrust/cub:cub/util_debug.cuh#L134-L141](https://github.com/thrust/cub/blob/8bdfee73970eac30c2c776956b5aab3fc87b604e/cub/util_debug.cuh#L134-L141)) that was introduced for Clang 4.0.0 (72fefca8). It's probably not relevant anymore for recent Clang versions. In fact, replacing `va_printf` by `printf` also removes the compiler error in ESPResSo:

```diff
--- cub/util_debug.cuh	2020-04-09 17:21:16.707835166 +0200
+++ cub/util_debug.cuh	2020-04-09 17:24:09.225871996 +0200
@@ -129,9 +129,9 @@
         #endif
             }
         #ifndef __CUDA_ARCH__
-            #define _CubLog(format, ...) va_printf(format,__VA_ARGS__);
+            #define _CubLog(format, ...) printf(format,__VA_ARGS__);
         #else
-            #define _CubLog(format, ...) va_printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, __VA_ARGS__);
+            #define _CubLog(format, ...) printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, __VA_ARGS__);
         #endif
     #endif
 #endif
```

I'm not really familiar with the thrust code, so take these two quick fixes with a grain of salt. The issue seems to have been fixed in thrust 1.9.5 because the compiler error is not reproducible with Clang++ 9.0 with CUDA 10.0 and thrust 1.9.5.

**Comment by allisonvacanti** [comment]
This was fixed with thrust/cub@a568ffa1fe061c20689934f119afd09beae820fd.
